### PR TITLE
chore: Handle .env variables

### PR DESCRIPTION
**Pull request by VancheeZze**
## Changes
There is a new mechanism of building `.env.development` and `.env.production` files using package-specific `env.config` scripts.

All environment variables are managed by an optional `bodiless.env.config.(js|ts)` file in the root of each package that needs to set or update environment variables. Default env variables can be found in `dist/generate-env-vars/getDefaults` in `@bodiless/gatsby-theme-bodiless` package and all env variables may be overwritten by setting them in `.env.site` in the root directory of the site.

#### About `bodiless.env.config.js` file
There is an optional `bodiless.env.config.js` file in a package root directory that is used to configure `.env` variables by extending passed env configuration during the site build. In order for it to work it must export an `async` `configure(defaultConfig, appEnv)` function. This function accepts 2 params:
* `defaultConfig` - this is an env configuration object with all current env variables.
* `appEnv` - current app build environment. **Note:** `appEnv` (`development` or `production`) actually describes whether we are in an edit environment or a static site build.
  * `production` - static site build (`gatsby build` == `production`)
  * `development` - edit environment (`gatsby develop` == `development`)

#### Example of `bodiless.env.config.js` file
```js
module.exports = {
  // We must export an async 'configure()' function. This function
  // is used by the main `generate-env-files` script.
  configure: async (defaultConfig, appEnv) => {
  // Here we define all env variables that the current package controls
  // based on the `appEnv`. We aslo recommend to have a 'default' value.
    const defaultEnvValues = {
      BODILESS_TAILWIND_THEME_ENABLED: '1',
      BODILESS_BACKEND_COMMIT_ENABLED: '0',
      BODILESS_BACKEND_SAVE_ENABLED: '1',
    };

    // Get current git branch
    const currentGitBranch = await getCurrentGitBranch();

    // Here we define all env variables that the current package controls
    // based on the `currentGitBranch`. We aslo recommend to have a 'default' value.
    const config = currentGitBranch === 'master'
      ? { default: { ...defaultEnvValues, BODILESS_BACKEND_SAVE_ENABLED: '0' } }
      : { default: { ...defaultEnvValues } }

    // There might be a cases when 'config' values are same for prod and dev.
    // In this case we may keep only the 'default' option in 'config' object.
    // `validNodeEnv` ensures that 'config' object has 'appEnv' option.
    // If this is not a case it will return a 'default' option.
    const validNodeEnv = val => Object.keys(config).includes(val);

  // Here we just merging a default config with the one provided by the package.
  // Note that it the same env variable occurs in both configs, the one that
  // defined in 'config[appEnv]' will take precedence over
  // the default one passed to the function.
    return {
      ...defaultConfig,
      ...validNodeEnv(appEnv)
        ? config[appEnv]
        : config.default,
    };
  },
};
```

The `bodiless.env.config.js` file is not limited to the plain objects and may execute any node logic to configure the env variables. The only rule is to return an extended version of the `defaultConfig` in the async `configure()` handler.

#### Dealing with precedence
Basically, there are 3 pieces of the `.env` configuration system:
* A default `getDefaults.ts` file with default env variables for bodiless sites defined in `@bodiless/gatsby-theme-bodiless`. 
* Optional `bodiless.env.config.js` files in the root of bodiless packages. These files take precedence over the default configuration. It may either extend the default configuration or overwrite certain variables. Note that there may be a `bodiless.env.config.js` in the site root folder which will take precedence over any other `.env` configurations.
* Site level `.env.site` file. This file is designed to take precedence over both: `getDefaults.ts` in `@bodiless/gatsby-theme-bodiless` and `bodiless.env.config.js` files. Also, there might be a need to dynamically set site-level env variables. You can do that by creating a `bodiless.env.config.js` file in the root folder of a site. This file will take precedence over any other env configuration.

## Test Instructions

1. **Test if `.env.production` and `.env.development` are generated when you run `npm run build`:**
   * Make sure there are no `.env.production` and `.env.development` files in `examples/test-site/`.
   * Run `npm run build` from either `./examples/test-site/` or `./` folder.
   * Make sure `.env.production` and `.env.development` files were generated during the build.
   * **Note** `npm run setup` currently does not trigger `npm run build` inside `/examples/` folder.

2. **Test if `.env.production` and `.env.development` are generated when you run `npm run start`:**
   * Make sure there are no `.env.production` and `.env.development` files in `examples/test-site/`.
   * Run `npm run start` from the root folder.
   * Make sure `.env.production` and `.env.development` files were generated and app starts with no errors.

3. **Test if `.env.site` overwrites the default env values:**
   * Update some env values in `./examples/test-site/.env.site` and generate new env variable by running `npm run build:env-vars` or `npm run build` from the `./examples/test-site/` directory. You can also run `npm run build` or `npm run start` from the root folder. It will regenerate env vars as well.
   * Make sure values you changed in `.env.site` are reflected ( overwritten ) in `.env.production` and `.env.development` files.

4. **Test if packages can overwrite the default env configuration:**
   * Create `bodiless.env.config.js` file in the root of any package.
   * You can use the `bodiless.env.config.js` file example above to configure or add some env variables.
   * Run `npm run setup`.
   * Generate new env variable by running `npm run build:env-vars` or `npm run build` from the `./examples/test-site/` directory. You can also run `npm run build` or `npm run start` from the root folder. It will regenerate env vars as well.
   * Make sure you see the `.env.production` and `.env.development` were updated according to the changes in `bodiless.env.config.js` you created.

5. **Test if `.env.site` overwrites the package env values from `bodiless.env.config.js`:**
   * Set some env values in `bodiless.env.config.js` in any package.
   * Overwrite the same env values in `./examples/test-site/.env.site` and generate new env variable by running `npm run build:env-vars` or `npm run build` from the `./examples/test-site/` directory. You can also run `npm run build` or `npm run start` from the root folder. It will regenerate env vars as well.
   * Make sure values you changed in `.env.site` are reflected ( overwritten ) in `.env.production` and `.env.development` files.

6. **Test if the app still works as expected:**
   * Run `npm run start`.
   * Make sure it starts with no errors.
   * Check if `Docs` button work.
   * Check if data is saved: change taco title and refresh the page - changes should be persisted.

## Related Issues
Fixes https://github.com/johnsonandjohnson/Bodiless-JS/issues/50

### CODE REVIEW

File: examples/starter/package.json [hunk 1]
```diff
@@ -59,11 +60,10 @@
     "start": "run-p -r dev-backend dev-frontend",
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/starter/package.json]: Done. Added `build:env-vars` to prestart.

File: examples/test-site/package.json [hunk 2]
```diff
@@ -53,34 +53,31 @@
...
-    "build:css": "tailwind build ./src/components/index.tailwind.css -c ./tailwind.config.js -o ./src/components/index.css",
+    "build:css": "node scripts/build-css.js",
     "build:doc": "bl-docs-build",
     "build:scripts": "tsc -p ./tsconfig.json",
     "check": "tsc -p ./tsconfig.check.json",
     "clean": "rimraf ./public/* ./.cache",
     "dev-backend": "node ./node_modules/@bodiless/backend/src/server.js",
     "dev-frontend": "gatsby develop -p 8005",
     "serve": "gatsby serve",
-    "prestart": "npm run build:doc",
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/test-site/package.json]: Updated prestart script

File: examples/test-site/package.json [hunk 3]
```diff
@@ -53,34 +53,31 @@
...
+    "prebuild:watch": "npm run build:doc",
     "start": "npm run build:watch",
     "generate-env-vars": "node scripts/generate-env-vars.js",
-    "generate-env-var:overwrite": "node scripts/generate-env-vars.js --overwrite",
-    "docs:build": "bl-docs-build",
-    "docs:serve": "docsify serve ./doc",
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/test-site/package.json]: Returned `docs:serve`

File: examples/starter/package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/starter/package.json]: Resolved. Added it to prestart

File: examples/starter/scripts/generate-env-vars.js [hunk 4]
```diff
@@ -12,40 +12,81 @@
...
-  }
-  const targetExists = fs.existsSync(target);
-  if (targetExists && !shouldOverwrite) {
-    logNotice(`environment variable file ${target} already exists. try using --overwrite if you want to overwrite it`);
-    return;
+const asyncGlob = util.promisify(glob);
+const readFile = util.promisify(fs.readFile);
+const writeFile = util.promisify(fs.writeFile);
+
+const DEFAULT_ENV_PATH = path.resolve('./node_modules/@bodiless/gatsby-theme-bodiless/.env.default');
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/starter/scripts/generate-env-vars.js]: Done. It is now moved to gatsby-theme-bodiless and has defaults included.

File: examples/starter/package.json [hunk 5]
```diff
@@ -31,6 +31,7 @@
     "dotenv": "^8.2.0",
     "express": "^4.17.1",
     "gatsby": "^2.9.11",
+    "glob": "^7.1.6",
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/starter/package.json]: Removed

File: examples/test-site/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/test-site/package.json]: Updated prestart script to build both: docs and env-vars

File: packages/gatsby-theme-bodiless/.env.default [hunk 6]
```diff
@@ -0,0 +1,5 @@
+BODILESS_BACKEND_DATA_FILE_PATH='src/data'
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/gatsby-theme-bodiless/.env.default]: Done

File: packages/gatsby-theme-bodiless/env.config.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      default: {
+        BODILESS_BACKEND_COMMIT_ENABLED: '0',
+        BODILESS_BACKEND_SAVE_ENABLED: '1',
+      },
+    };
+
+    const validNodeEnv = val => Object.keys(config).includes(val);
+    const isChangeset = branchName => branchName.startsWith('test/') || branchName.startsWith('changeset/');
+
+    const branch = await gitCurrentBranch();
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/gatsby-theme-bodiless/env.config.js]: Refactored. It now handles the "detached head" state and may be extended with additional checks.

File: packages/bodiless-documentation/doc/About/GettingStarted.md [hunk 8]
```diff
@@ -63,6 +63,56 @@ npm run setup
 ```
 > Note: don't run `npm install` at package root unless you are trying to update dependencies.
 
+### Environment Variables
+All environment variables are managed by an optional `env.config` file in the root of each package that needs to set or update environment variables. Default env variables can be found in `.env.default` in `@bodiless/gatsby-theme-bodiless` package and all env variables may be overwritten by setting them in `.env.site` in the root directory of the site.
+
+#### About `env.config` file:
+There is an optional `env.config` file in a package root directory that is used to configure `.env` variables by extending passed env configuration during the site build. In order for it to work it must export an `async` `configure(defaultConfig, nodeEnv)` function. This function accepts 2 params:
+* `defaultConfig` - this is an env configuration object with all current env variables.
+* `nodeEnv` - current node environment ( `production` or `development` ).
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/bodiless-documentation/doc/About/GettingStarted.md]: Added a note to the documentstion

File: packages/gatsby-theme-bodiless/package.json [hunk 9]
```diff
@@ -17,7 +17,8 @@
     "gatsby-config.js",
     "gatsby-node.js",
     "queries.jsx",
-    "Logger.js"
+    "Logger.js",
+    "env.config.js"
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/gatsby-theme-bodiless/package.json]: Removed

File: examples/starter/scripts/generate-env-vars.js [hunk 10]
```diff
@@ -12,40 +12,81 @@
...
+        ...await require(path.resolve(envConfigPath)).configure(agregatedEnvConfig, nodeEnv),
+      };
+    }
+
+    return agregatedEnvConfig;
+  }, Promise.resolve(defaultConfig));
+};
+
+const init = async () => {
+  const defaultEnvConfig = await envToJson(DEFAULT_ENV_PATH);
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/starter/scripts/generate-env-vars.js]: Done. Defaults are now moved into the generate-env-vars script.

File: packages/gatsby-theme-bodiless/package.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/gatsby-theme-bodiless/package.json]: Removed

File: examples/test-site/scripts/generate-env-vars.js [hunk 11]
```diff
@@ -12,40 +12,81 @@
  * limitations under the License.
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/test-site/scripts/generate-env-vars.js]: Resolved as well

File: packages/bodiless-backend/env.config.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
+/**
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/bodiless-backend/env.config.js]: Removed now

File: examples/starter/scripts/generate-env-vars.js [hunk 13]
```diff
@@ -12,40 +12,81 @@
...
-generateEnvironmentVariables('.env.default', '.env.development');
-generateEnvironmentVariables('.env.default', '.env.production');
-logNotice('environment variables generation finished');
+
+  return siteEnvFile;
+};
+
+const getBodilessEnvConfig = async (defaultConfig, nodeEnv) => {
+  const ignore = ['**/node_modules/@bodiless/**/node_modules/**'];
+  const globPattern = '**/node_modules/@bodiless/**/env.config.{js,ts}';
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply examples/starter/scripts/generate-env-vars.js]: Resolved. Not using a mechanism similar to `locateFiles` to find all `bodiless.env.config` files

File: packages/bodiless-core/env.config.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
+/**
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/bodiless-core/env.config.js]: Removed `env.config.js ` from `bodiless-core`

File: packages/bodiless-documentation/doc/About/GettingStarted.md [hunk 15]
```diff
@@ -63,6 +63,56 @@ npm run setup
...
+
+#### Example of `env.config` file:
+```js
+module.exports = {
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/bodiless-documentation/doc/About/GettingStarted.md]: Updated the example as well

File: packages/bodiless-documentation/doc/About/GettingStarted.md [hunk 16]
```diff
@@ -63,6 +63,56 @@ npm run setup
 ```
 > Note: don't run `npm install` at package root unless you are trying to update dependencies.
 
+### Environment Variables
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/bodiless-documentation/doc/About/GettingStarted.md]: Restructured.

File: packages/gatsby-theme-bodiless/env.config.js [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+    if (nodeEnv !== 'production' && isChangeset(branch)) {
+      nodeEnv = 'changeset';
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/gatsby-theme-bodiless/env.config.js]: Renamed it to the `appEnv` everywhere.

**Comment by wodenx** [comment]
@VancheeZze I question whether "full regression testing" is required.  It seems to me sufficient to test that all env vars are set properly in .env.development and .env.production after a build, no?  cc @TatsianaMiraniuk @dewen

**Comment by VancheeZze** [comment]
> @VancheeZze I question whether "full regression testing" is required. It seems to me sufficient to test that all env vars are set properly in .env.development and .env.production after a build, no? cc @TatsianaMiraniuk @dewen

It might be enough, but I think it would also be beneficial to do at least a quick smoke testing ( building the app, making sure it is launching fine and the basic functionality controlled by env vars are working fine. Such as docs url for example or saving the data  )

File: examples/starter/package.json [hunk 18]
```diff
@@ -46,24 +46,23 @@
     "url": "https://github.com/johnsonandjohnson/bodiless-js.git"
   },
   "scripts": {
-    "build": "npm-run-all build:css build:lib build:doc",
+    "build": "npm-run-all build:env-vars build:lib build:doc",
```
Comment (wodenx) [inline examples/starter/package.json]: Shouldn't we still be running build:css here?

Comment (VancheeZze) [reply examples/starter/package.json]: I think `build:css` is now moved to `prepare` script: `"prepare": "npm run build:css"` which is executed before the build anyway.

File: packages/gatsby-theme-bodiless/package.json [hunk 19]
```diff
@@ -5,8 +5,12 @@
   "author": "Chris Oden <coden@its.jnj.com>",
   "homepage": "",
   "license": "Apache-2.0",
+  "bin": {
+    "generate-env-vars": "./bin/generate-env-vars"
+  },
   "directories": {
     "lib": "lib",
+    "dist": "dist",
```
[the remark this answers is not in the document]
Comment (VancheeZze) [reply packages/gatsby-theme-bodiless/package.json]: Done

**Pull request merged**